[XLA] Add a new XLA mode: XLALite

nouiz opened the pull request:
Add 2 new XLA flags:
- `TF_XLA_FLAGS=--tf_xla_ops_to_cluster=[FUSIBLE,...]`
- `TF_XLA_FLAGS=--tf_xla_auto_jit=fusible`
    This is a shortcut to `TF_XLA_FLAGS=--tf_xla_ops_to_cluster=FUSIBLE TF_XLA_FLAGS=--tf_xla_auto_jit=1`


This enables XLA but only for a subset of TF operations that XLA know how to fuse together. This allows using XLA operations fusion capabilities while removing some of the current XLA slow down case. In most cases where XLA is slower than TF classic, XLA isn't slower the TF classic. In some cases where XLA give speed up vs TF classic, XLALite give a good part of that speed up.

The flag tf_xla_supported_ops can accept TF operation names and/or predefined groups of operation. The group FUSIBLE  includes all the groups defined. Multiple value can be passed by separating them by comma.

nouiz committed:
Small comment fix

nouiz committed:
Add tf_xla_supported_nodes flags to limits nodes XLA consider.

nouiz committed:
More Ops in the whitelist category.

nouiz committed:
Add more XLA whitelist shortcut

nouiz committed:
Better user error detection, less verbose and better error message.

nouiz committed:
Add an EXTRA category

nouiz committed:
Add the flag value TF_XLA_FLAGS=--tf_xla_auto_jit=fusible to enable XLA and the tf_xla_supported_nodes=FUSIBLE flag. This make using that mode easier.

nouiz committed:
Update docmentation of the new flag

nouiz committed:
Use the absl containers.

nouiz committed:
Move code to a function to make the code more clear.

nouiz committed:
Rename the new flags to tf_xla_supported_ops

nouiz committed:
Typo and add comment.

nouiz committed:
Replace a series of if with a table. Make code simpler to understand.

nouiz committed:
Small update (typo, clang-format, const)

nouiz committed:
XLALite, put BiasAddGrad in the right section and add the missing BiasAdd operation.

nouiz committed:
Code formating

nouiz committed:
Repair the XLALite flag shortcut since the rebase

nouiz committed:
clang-format

nouiz committed:
Add TopKV2 to XLALite whitelist. XLA version is much faster the TF.

thomasjoerg commented on tensorflow/compiler/jit/mark_for_compilation_pass.cc:
```diff
@@ -1076,6 +1076,86 @@ StatusOr<bool> IsIdentityDrivingConstsInLoop(Node* node) {
   return true;
 }
 
+const absl::flat_hash_map<string, std::vector<string>> whitelist_table = {
```
What invariants must hold for this list? Or phrased differently, at what points does this list to be updated? Is it meant to be complete, i.e. all point-wise operations supported by XLA:GPU are listed here? In you experiments, did you find subsets of `FUSIBLE` useful?

nouiz replied:
Every times the TF-XLA bridge support a new TF operation, we need to check if we want to support it by default or not. So we should document that somewhere to help this not to be forgotten. Any idea where we should document that?

I didn't used the shortcut much. What I used is FUSIBLE plus other operation that I wasn't sure if we want to include or not. But if someone want to start to play more with it, I think the shortcut would be useful.

To make this list, I passed over the TF operations that the bridge knows and selected those that I was sure what they do and that I was sure XLA could fuse them. I could have missed some. In 2 benchmarks, I found some that could maybe fused that I didn't include. I timed when including them and it slowed down XLALite compared to not having them. So I didn't included them. They where:
```   {"AssignVariableOp", "AssignAddVariableOp", "AssignSubVariableOp",
   "ReadVariableOp", "VarIsInitializedOp", "VariableShape",
   "ResourceApplyCenteredRMSProp", "ResourceApplyRMSProp",
   "ResourceScatterAdd", "ResourceApplyAdam"}```

thomasjoerg replied:
@sanjoy Can you recommend a good spot for this documentation? Do you have an opinion regarding the TF Ops Frederic did not include?

sanjoy replied:
As for the documentation I would vote for creating a unit test that checks that all non-whitelisted tf2xla ops are from a fixed set.  See [resource_operation_table_test.cc](https://github.com/tensorflow/tensorflow/blob/3039a4694e22674b502257ae34b0a5b614a631f3/tensorflow/compiler/tf2xla/resource_operation_table_test.cc#L1).

> Do you have an opinion regarding the TF Ops Frederic did not include?

Right it isn't totally obvious which ops were included.  I think we need some comments to describe the "format" of `whitelist_table`.

nouiz replied:
Good idea to have a test that force to have all XLA/TF operations registered to be whiteliste or blacklisted. It will guaranty that we make a decision when new operations are supported.
I added this test.

nouiz committed:
Remove some duplicate names.

nouiz committed:
Better documentation of the new parameter

nouiz committed:
Rename tf_xla_supported_ops to tf_xla_ops_to_cluster

nouiz commented on tensorflow/compiler/jit/flags.cc:
```diff
@@ -78,6 +89,21 @@ void AppendMarkForCompilationPassFlagsInternal(std::vector<Flag>* flag_list) {
+           "If multiple, separate them with commas. Shortcuts: "
+           " PW: All point-wise operations."
+           " RED: All reduction operations."
+           " SMALL: Mixed small operations."
```
Small is like Shape, Rank, Range and a few others I didn't know in which category to put. I could split that section into 'SMALL' and MIXED. Do you have a better idea of how to split that?

thomasjoerg replied:
Since you said you rarely used the subcategories of FUSIBLE anyway, I wouldn't split them further. I was just wondering in what way the TF ops are SMALL. If this is a category for ops that just wouldn't fit anywhere else, I'd call it MISC to make this obvious.

nouiz replied:
Done

nouiz commented on tensorflow/compiler/jit/flags.cc:
```diff
@@ -65,7 +74,9 @@ void AppendMarkForCompilationPassFlagsInternal(std::vector<Flag>* flag_list) {
-           "things very likely to be improved; 2 = on for everything.  "
+           "things very likely to be improved; 2 = on for everything; "
+           "fusible = only for Tensorflow operations that XLA knows how to "
+           "fuse.  "
```
done

cheshire replied:
Not done? Also should this be under a different flag if this can change in unpredictable ways?

nouiz commented on tensorflow/compiler/jit/flags.h:
```diff
@@ -55,6 +55,9 @@ struct MarkForCompilationPassFlags {
   int32 tf_xla_max_cluster_size;
 
+  // If non-empty, limit XLA clustering to the following TF operations.
+  string tf_xla_supported_ops;
```
done

nouiz commented on tensorflow/compiler/jit/flags.cc:
```diff
@@ -78,6 +89,21 @@ void AppendMarkForCompilationPassFlagsInternal(std::vector<Flag>* flag_list) {
+           "(LRN, LRNGrad)."
+           " BN: TF FusedBatchNorm* operations."
+           " FUSIBLE: All TF operations that XLA can fuse (All the above). "
+           "You can also put any TF operation name."),
```
I added an example. I'm not sure what you mean by fully-qualified name? I think TF operation have just one unique name like Add, Matmul, Sum,... It is those that should be used in this version.

cheshire commented on tensorflow/compiler/jit/mark_for_compilation_pass.cc:
```diff
@@ -1076,6 +1076,86 @@ StatusOr<bool> IsIdentityDrivingConstsInLoop(Node* node) {
+
+std::unique_ptr<absl::flat_hash_set<string>> GetWhitelist() {
+  MarkForCompilationPassFlags* flags = GetMarkForCompilationPassFlags();
+  auto whitelist = absl::WrapUnique(new absl::flat_hash_set<string>());
```
make_unique?

cheshire commented on tensorflow/compiler/jit/mark_for_compilation_pass.cc:
```diff
@@ -1076,6 +1076,86 @@ StatusOr<bool> IsIdentityDrivingConstsInLoop(Node* node) {
+
+  if (VLOG_IS_ON(2) && whitelist->size() > 0) {
+    std::vector<string> vwhitelist(whitelist->begin(), whitelist->end());
+    std::sort(vwhitelist.begin(), vwhitelist.end());
```
absl::c_sort

nouiz replied:
done.

cheshire commented on tensorflow/compiler/jit/mark_for_compilation_pass.cc:
```diff
@@ -1076,6 +1076,86 @@ StatusOr<bool> IsIdentityDrivingConstsInLoop(Node* node) {
+    }
+  }
+
+  if (VLOG_IS_ON(2) && whitelist->size() > 0) {
```
!whitelist->empty()

cheshire commented on tensorflow/compiler/jit/mark_for_compilation_pass.cc:
```diff
@@ -1076,6 +1076,86 @@ StatusOr<bool> IsIdentityDrivingConstsInLoop(Node* node) {
+    } else if (whitelist_table.contains(s)) {
+      auto v = whitelist_table.at(s);
+      whitelist->insert(v.begin(), v.end());
+    } else if (s.size() > 0) {
```
!s.empty()

cheshire commented on tensorflow/compiler/jit/mark_for_compilation_pass.cc:
```diff
@@ -1076,6 +1076,86 @@ StatusOr<bool> IsIdentityDrivingConstsInLoop(Node* node) {
+      whitelist->insert(v.begin(), v.end());
+    } else if (s.size() > 0) {
+      // Should be a user provided TF operation.
+      whitelist->insert(string(s));
```
should we VLOG(5) here or something to avoid misspellings?

cheshire commented on tensorflow/compiler/jit/mark_for_compilation_pass.cc:
```diff
@@ -1076,6 +1076,86 @@ StatusOr<bool> IsIdentityDrivingConstsInLoop(Node* node) {
+    }
+  }
+
+  if (VLOG_IS_ON(2) && whitelist->size() > 0) {
```
Why not just sort on all codepaths, then this entire branch becomes `VLOG(2) << ...`

nouiz replied:
To remove the time for the sort when we do not print.
I tried to change the container to the sorted container absl::btree_set to remove the need for a sort. But btree_set isn't available in my version of absl. I'm not sure it is a good reason to bump TF absl required version and never changed TF dependency version.

If I put a VLOG(2) on all code path, this will make the output verbose even when this feature isn't used. Are you suggesting doing that? I wouldn't have a tendency of adding extra useless verbose output here. Do you see value to always print it even when not used?

cheshire commented on tensorflow/compiler/jit/mark_for_compilation_pass.cc:
```diff
@@ -1108,6 +1187,19 @@ Status MarkForCompilationPassImpl::FindCompilationCandidates() {
 
   VLOG(2) << "sorted_nodes.size() = " << sorted_nodes.size();
 
+  auto whitelist = GetWhitelist();
```
nitpicking, but for a small value copied once I would just avoid the unique_ptr altogether

cheshire commented on tensorflow/compiler/jit/mark_for_compilation_pass.cc:
```diff
@@ -1108,6 +1187,19 @@ Status MarkForCompilationPassImpl::FindCompilationCandidates() {
 
+  auto whitelist = GetWhitelist();
+
+  auto vall_ops = XlaOpRegistry::GetAllRegisteredOps();
```
Explicit type would be useful here

cheshire commented on tensorflow/compiler/jit/mark_for_compilation_pass.cc:
```diff
@@ -1108,6 +1187,19 @@ Status MarkForCompilationPassImpl::FindCompilationCandidates() {
+  auto vall_ops = XlaOpRegistry::GetAllRegisteredOps();
+  absl::flat_hash_set<string> all_ops(vall_ops.begin(), vall_ops.end());
+  // Check that user's provided TF operation really exists.
+  for (auto s = whitelist->begin(); s != whitelist->end(); ++s) {
```
for-each loop instead?

nouiz replied:
done

nouiz committed:
Added warning about an experimental feature and code simplification.

nouiz committed:
Remove the unique ptr to simplify the code.

nouiz committed:
Small code simplification.

cheshire commented on tensorflow/compiler/jit/flags.cc:
```diff
@@ -65,7 +74,9 @@ void AppendMarkForCompilationPassFlagsInternal(std::vector<Flag>* flag_list) {
            "GPU devices.  0 = use ConfigProto setting; -1 = off; 1 = on for "
-           "things very likely to be improved; 2 = on for everything.  "
+           "things very likely to be improved; 2 = on for everything; "
+           "(experimental) fusible = only for Tensorflow operations that XLA "
```
Do we actually need a separate mode here?
I don't like "fusible" too much because it's not consistent with previously numeric values, and I also wonder why is it not sufficient to set `--tf_xla_auto_jit=2 --tf_xla_ops_to_cluster=FUSIBLE` (we can also have `--tf_xla_ops_to_cluster=FUSIBLE` imply `--tf_xla_auto_jit=2`). 

I would also like this mode to be a separate flag for now.

nouiz replied:
TF already have tons of flags. This make them more difficult to discover and be used by users. Reusing existing flags help user find, use them and remember them. This is why we prefer to have a shortcut with existing flags. Changing the name of the new flag value to something is not a problem. Any suggestion? "lite"(as the project is called XLALite)?
This option heavily influence the auto_jit, so it is highly related to this parameter. Ideally, I would have created a new number like 1 and bumped the current value 1 and 2. But this would break the interface which isn't great.

Currently, --tf_xla_auto_jit=fusible imply --tf_xla_auto_jit=1, not =2. It is  not clear what the difference is between them for me and I didn't found the code which make a difference in which nodes get selected or not. All benchmarks have been done with 1. Is it a problem to change the value from 1 to 2?

thomasjoerg replied:
I prefer using `--tf_xla_auto_jit` as Frédéric proposed.

George, Sanjoy, and I have a meeting tomorrow and we'll put this question on the agenda. We'll get back to you tomorrow, Frédéric.

cheshire replied:
OK!

cheshire commented on tensorflow/compiler/jit/mark_for_compilation_pass.cc:
```diff
@@ -1076,6 +1076,86 @@ StatusOr<bool> IsIdentityDrivingConstsInLoop(Node* node) {
   return true;
 }
 
+const absl::flat_hash_map<string, std::vector<string>> whitelist_table = {
```
This looks like a non-trivially-destructible global, Google style guide prohibits those (cf. https://google.github.io/styleguide/cppguide.html#Static_and_Global_Variables)

nouiz replied:
Fixed.

cheshire commented:
@thomasjoerg the issue with the non-trivially destructible global remains, it will break the build.

sanjoy commented on tensorflow/compiler/jit/mark_for_compilation_pass.cc:
```diff
@@ -1076,6 +1076,86 @@ StatusOr<bool> IsIdentityDrivingConstsInLoop(Node* node) {
   return true;
 }
 
+const absl::flat_hash_map<string, std::vector<string>> whitelist_table = {
```
Please document the format of this table.

We also don't allow non-trivial global destructors because they don't play well with multi-threading.  I think a better phrasing is:

```
absl::flat_hash_map<string, std::vector<string>> *CreateWhitelist() {
  absl::flat_hash_map<string, std::vector<string>>* result = new ...;
  // Use explicit code to populate "result", possibly with comments.
}

const absl::flat_hash_map<string, std::vector<string>>& GetOrCreateWhitelist() {
  static absl::flat_hash_map<string, std::vector<string>>* whitelist = CreateWhitelist();
  return *whitelist;
}
```

nouiz replied:
I fixed the non-trivial global destructors.
I also added documentation of the format. If you wanted more then this, just tell me.

cheshire replied:
I think it's better to have `static` in `CreateWhitelist` in order not accidentally create a leak if someone else calls it, also it seems this is how it is usually done in XLA, e.g. https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/xla/service/transfer_manager.cc#L41

nouiz committed:
Convert non-trivial global destructors to local static variable.

nouiz committed:
Rename a category

nouiz committed:
Add test to make sure all XLA support operation are in the XLALite whitelist or blacklist.

cheshire commented on tensorflow/compiler/jit/mark_for_compilation_pass.cc:
```diff
@@ -1714,7 +1761,300 @@ Status MarkForCompilationPass::RunForTest(
 
+absl::flat_hash_map<string, std::vector<string>>* CreateWhitelist() {
+  // Table format: category name: {list of TF operations in that category}
+  absl::flat_hash_map<string, std::vector<string>>* result =
```
Should this be static? At the moment this seems to leak on every invocation.

nouiz replied:
I modified it. See other comment: https://github.com/tensorflow/tensorflow/pull/34655#discussion_r353833354

nouiz committed:
Move the static inside the function to be safer.

nouiz commented:
I can't reply to the comment, so replying here.
I changed the code (and renamed the function name) to be safer to potential future changes as suggested.

thomasjoerg commented:
@rthadur Did CopyBara get stuck somehow? It's been two days since the PR was approved. Can you kick CopyBara?

cheshire commented:
@thomasjoerg we can also force run copybara by adding the `kokoro:force-run` tag, let me try to do it.

thomasjoerg commented:
> @thomasjoerg we can also force run copybara by adding the `kokoro:force-run` tag, let me try to do it.

@cheshire Frocing a Kokoro did not do the trick. I imported the PR manually.

nouiz committed:
Fix compilation crash for the new XLA test since the interface change.

nouiz commented:
Note, I forgot to update the interface change in my test. So my test currently broke.
Why the CI didn't failed? Do they run XLA tests?
I just pushed a fix. I saw somewhere that tf nightly build are broken. Could this be related?

nouiz committed:
Print the name that should be used to enable it. This
 make it easier to find which operation. For example, the nodename can be AvgPool2d, while the TF operation is AvgPool.

nouiz commented:
I made a new PR with the last commit that was missing:
https://github.com/tensorflow/tensorflow/pull/35263
